# Finding our best-performing salespeople and products

## Introduction

**Business Context.** You work for AdventureWorks, a company that sells outdoor sporting equipment. The company has many different locations and has been recording the sales of different locations on various products. You, their new data scientist, have been tasked with the question: **"What are our best products and salespeople and how can use this information to improve our overall performance?"**

You have been given access to the relevant data files with documentation from the IT department. Your job is to extract meaningful insights from these data files to help increase sales. First, you will look at the best products and try to see how different products behave in different categories. Second, you will analyze the best salespeople to see if the commission percentage motivates them to sell more.

**Business Problem.** Your task is to **construct a database from the provided CSV files and then write queries in SQL to carry out the requested analysis**.

**Analytical Context.** You are given the data (stored in the ```data/csvs``` folder) as a set of separate CSV files, each one representing a table. You will build a new PostgreSQL database from these files using AWS RDS.

The company has been pretty vague about how they expect you to extract insights, but you have come up with the following plan of attack:

1. Create the database and ensure you can run basic queries against it
2. Look at how product ratings and total sales are related
3. See how products sell in different subcategories (bikes, helmets, socks, etc.)
4. Calculate which salespeople have performed the best in the past year
5. Seeing if total sales are correlated with their commission percentage

Of course, this is only your initial plan. As you explore the database, your strategy will change.

## Setting up AWS

In this case, we'll assume that the company has given you an entry-level laptop, which is not capable of running a PostgreSQL server locally. Therefore, you should set up a cloud database, connect to it from `psql`, and run the analysis via the `psql` or directly from the notebook.

### Question:

Repeat the steps in Case 12.3 to create a new RDS instance with a PostgreSQL database.

In [6]:
'''RDS instance name: extended-case-3'''

'RDS instance name: extended-case-3'

## Overview of the data

The data for the case is contained in the ```./data/csvs``` directory; specifically, it is the ```AdventureWorks``` sample data provided by Microsoft. We will be focusing on the Sales and Production categories. Complete documentation for the original data (of which you have only a subset) can be found [here](https://dataedo.com/download/AdventureWorks.pdf). 

**Product Tables:**
* **Product**: one row per product that the company sells
* **ProductReview**: one row per rating and review left by customers
* **ProductModelProductDescriptionCulture**: a link between products and their longer descriptions also indicating a "culture" - which language and region the product is for
* **ProductDescription**: a longer description of each product, for a specific region
* **ProductCategory**: the broad categories that products fit into
* **ProductSubCategory**: the narrower subcategories that products fit into

**Sales Tables:**
* **SalesPerson**: one row per salesperson, including information on their commission and performance
* **SalesOrderHeader**: one row per sale summarizing the sale
* **SalesOrderDetail**: many rows per sale, detailing each product that forms part of the sale
* **SalesTerritory**: the different territories where products are sold, including performance

**Region Tables:**
* **CountryRegionCurrency**: the currency used by each region
* **CurrencyRate**: the average and closing exchange rates for each currency compared to the USD

## Setting up `ipython-sql` and `pgspecial`

Jupyter notebook is usually used to run Python code, but with an add-on it can run SQL directly against a database too. Install the extensions `ipython-sql` and `pgspecial` through `pip` (you may have to restart the notebook after doing this) and create the database `adventureworks`:

In [72]:
# !pip3 install ipython-sql pgspecial

Now load the sql add-on and connect to the database as follows. You'll need to change the username (`postgres`), password (`mysecretpassword`), host (`localhost`), and database name (`postgres`) to what you used when setting up your RDS instance:

**Using these commands in the psql shell to access the database in the RDS instance**

psql -h extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com -U postgres

password: zVM5Y5iSn9vAK0To51WE

\c adventureworks

In [73]:
%reload_ext sql
%sql postgresql://postgres:zVM5Y5iSn9vAK0To51WE@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks

'Connected: postgres@adventureworks'

You should now be able to run SQL directly from any Jupyter notebook cell by starting the cell with a line that states `%%sql`. For example (once you have a database with some tables, which we'll only create later):

```sql
%%sql

SELECT * FROM product LIMIT 10;
```

**Note:** Unlike `pandas` which automatically truncates output for large DataFrames, the SQL plug-in gives you exactly what you ask for. If you do a `SELECT * FROM` a table with a million rows and no `LIMIT` clause, it'll output all million rows and probably freeze your notebook. It's good practice to always use a `LIMIT` clause even when it's not needed to avoid any mishaps.

## Creating the database and adding the tables

Now, let's create a database called `adventuretime`. (If you do this through the notebook, you'll have to add the line `end;` before your `create database` command as the add-on runs everything in transactions).

You'll need to add a table for each of the CSV files. Spend some time looking at the different CSV files and getting used to how they reference each other and what headers they create. Then, you'll need to write an appropriate `CREATE TABLE` command with appropriate types. You can figure out the types by inspecting the CSV files and/or referencing the documentation.

### Exercise 1:

Write all of the commands that you need to

* Create the database
* Create the tables
* Import the data from the CSVs

**Hint:** As an example, to add data for the `salesperson` table, you would use the following commands:

1. Create table (can be run from Jupyter Notebook or the `psql` command line interface):
```sql
CREATE TABLE salesperson (
    businessentityid INTEGER,
    territoryid INTEGER,
    salesquota INTEGER,
    bonus INTEGER,
    commissionpct FLOAT,
    salesytd FLOAT,
    saleslastyear FLOAT,
    rowguid TEXT,
    modifieddate DATE
    );
```

2. Copy data (has to be run from the `psql` shell):

```sql
\copy salesperson FROM 'data/csvs/salesperson.csv' with (format CSV, header true, delimiter ',');
```

Database created from the shell using:

create database adventureworks;

\q


**Tables created and data copied from the psql shell with following commands:**

CREATE TABLE BusinessEntity(
    BusinessEntityID SERIAL, --  NOT FOR REPLICATION
    rowguid VARCHAR(50) NOT NULL  , -- ROWGUIDCOL
    ModifiedDate TIMESTAMP NOT NULL 
  );

CREATE TABLE Person(
    BusinessEntityID INT NOT NULL,
    PersonType char(2) NOT NULL,
    NameStyle boolean NOT NULL CONSTRAINT "DF_Person_boolean" DEFAULT (false),
    Title varchar(8) NULL,
    FirstName  varchar(50) NOT NULL,
    MiddleName  varchar(50) NULL,
    LastName  varchar(50) NOT NULL,
    Suffix varchar(10) NULL,
    EmailPromotion INT NOT NULL CONSTRAINT "DF_Person_EmailPromotion" DEFAULT (0),
    AdditionalContactInfo XML NULL, -- XML("AdditionalContactInfoSchemaCollection"),
    Demographics XML NULL, -- XML("IndividualSurveySchemaCollection"),
    rowguid varchar(50) NOT NULL, -- ROWGUIDCOL
    ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_Person_ModifiedDate" DEFAULT (NOW()),
    CONSTRAINT "CK_Person_EmailPromotion" CHECK (EmailPromotion BETWEEN 0 AND 2),
    CONSTRAINT "CK_Person_PersonType" CHECK (PersonType IS NULL OR UPPER(PersonType) IN ('SC', 'VC', 'IN', 'EM', 'SP', 'GC'))
  );
  
CREATE TABLE ProductSubcategory(
  ProductSubcategoryID SERIAL NOT NULL, -- int
  ProductCategoryID INT NOT NULL,
  Name varchar(50) NOT NULL,
  rowguid varchar(50) NOT NULL, -- ROWGUIDCOL
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_ProductSubcategory_ModifiedDate" DEFAULT (NOW())
);
CREATE TABLE ProductModel(
  ProductModelID SERIAL NOT NULL, -- int
  Name varchar(50) NOT NULL,
  CatalogDescription XML NULL, -- XML(ProductDescriptionSchemaCollection)
  Instructions XML NULL, -- XML(ManuInstructionsSchemaCollection)
  rowguid varchar(50) NOT NULL , -- ROWGUIDCOL
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_ProductModel_ModifiedDate" DEFAULT (NOW())
);

CREATE TABLE Product(
  ProductID SERIAL NOT NULL, -- int
  Name varchar(50) NOT NULL,
  ProductNumber varchar(25) NOT NULL,
  MakeFlag boolean NOT NULL CONSTRAINT "DF_Product_MakeFlag" DEFAULT (true),
  FinishedGoodsFlag boolean NOT NULL CONSTRAINT "DF_Product_FinishedGoodsFlag" DEFAULT (true),
  Color varchar(15) NULL,
  SafetyStockLevel smallint NOT NULL,
  ReorderPoint smallint NOT NULL,
  StandardCost numeric NOT NULL, -- money
  ListPrice numeric NOT NULL, -- money
  Size varchar(5) NULL,
  SizeUnitMeasureCode char(3) NULL,
  WeightUnitMeasureCode char(3) NULL,
  Weight decimal(8, 2) NULL,
  DaysToManufacture INT NOT NULL,
  ProductLine char(2) NULL,
  Class char(2) NULL,
  Style char(2) NULL,
  ProductSubcategoryID INT NULL,
  ProductModelID INT NULL,
  SellStartDate TIMESTAMP NOT NULL,
  SellEndDate TIMESTAMP NULL,
  DiscontinuedDate TIMESTAMP NULL,
  rowguid varchar(50) NOT NULL , -- ROWGUIDCOL
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_Product_ModifiedDate" DEFAULT (NOW()),
  CONSTRAINT "CK_Product_SafetyStockLevel" CHECK (SafetyStockLevel > 0),
  CONSTRAINT "CK_Product_ReorderPoint" CHECK (ReorderPoint > 0),
  CONSTRAINT "CK_Product_StandardCost" CHECK (StandardCost >= 0.00),
  CONSTRAINT "CK_Product_ListPrice" CHECK (ListPrice >= 0.00),
  CONSTRAINT "CK_Product_Weight" CHECK (Weight > 0.00),
  CONSTRAINT "CK_Product_DaysToManufacture" CHECK (DaysToManufacture >= 0),
  CONSTRAINT "CK_Product_ProductLine" CHECK (UPPER(ProductLine) IN ('S', 'T', 'M', 'R') OR ProductLine IS NULL),
  CONSTRAINT "CK_Product_Class" CHECK (UPPER(Class) IN ('L', 'M', 'H') OR Class IS NULL),
  CONSTRAINT "CK_Product_Style" CHECK (UPPER(Style) IN ('W', 'M', 'U') OR Style IS NULL),
  CONSTRAINT "CK_Product_SellEndDate" CHECK ((SellEndDate >= SellStartDate) OR (SellEndDate IS NULL))
);

CREATE TABLE ProductDescription(
  ProductDescriptionID SERIAL NOT NULL, -- int
  Description varchar(400) NOT NULL,
  rowguid varchar(50) NOT NULL, -- ROWGUIDCOL
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_ProductDescription_ModifiedDate" DEFAULT (NOW())
);

CREATE TABLE ProductModelProductDescriptionCulture(
  ProductModelID INT NOT NULL,
  ProductDescriptionID INT NOT NULL,
  CultureID char(6) NOT NULL,
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_ProductModelProductDescriptionCulture_ModifiedDate" DEFAULT (NOW())
);

CREATE TABLE ProductReview(
  ProductReviewID SERIAL NOT NULL, -- int
  ProductID INT NOT NULL,
  ReviewerName varchar(50) NOT NULL,
  ReviewDate TIMESTAMP NOT NULL CONSTRAINT "DF_ProductReview_ReviewDate" DEFAULT (NOW()),
  EmailAddress varchar(50) NOT NULL,
  Rating INT NOT NULL,
  Comments varchar(3850),
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_ProductReview_ModifiedDate" DEFAULT (NOW()),
  CONSTRAINT "CK_ProductReview_Rating" CHECK (Rating BETWEEN 1 AND 5)
);

CREATE TABLE ProductReview(
  ProductReviewID SERIAL NOT NULL, -- int
  ProductID INT NOT NULL,
  ReviewerName varchar(50) NOT NULL,
  ReviewDate TIMESTAMP NOT NULL CONSTRAINT "DF_ProductReview_ReviewDate" DEFAULT (NOW()),
  EmailAddress varchar(50) NOT NULL,
  Rating INT NOT NULL,
  Comments varchar(3850),
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_ProductReview_ModifiedDate" DEFAULT (NOW()),
  CONSTRAINT "CK_ProductReview_Rating" CHECK (Rating BETWEEN 1 AND 5)
);

CREATE TABLE SalesOrderHeader(
  SalesOrderID SERIAL NOT NULL, --  NOT FOR REPLICATION -- int
  RevisionNumber smallint NOT NULL CONSTRAINT "DF_SalesOrderHeader_RevisionNumber" DEFAULT (0), -- tinyint
  OrderDate TIMESTAMP NOT NULL CONSTRAINT "DF_SalesOrderHeader_OrderDate" DEFAULT (NOW()),
  DueDate TIMESTAMP NOT NULL,
  ShipDate TIMESTAMP NULL,
  Status smallint NOT NULL CONSTRAINT "DF_SalesOrderHeader_Status" DEFAULT (1), -- tinyint
  OnlineOrderFlag boolean NOT NULL CONSTRAINT "DF_SalesOrderHeader_OnlineOrderFlag" DEFAULT (true),
  SalesOrderNumber VARCHAR(23), -- AS ISNULL(N'SO' + CONVERT(nvarchar(23), SalesOrderID), N'*** ERROR ***'),
  PurchaseOrderNumber VARCHAR(25) NULL,
  AccountNumber VARCHAR(25) NULL,
  CustomerID INT NOT NULL,
  SalesPersonID INT NULL,
  TerritoryID INT NULL,
  BillToAddressID INT NOT NULL,
  ShipToAddressID INT NOT NULL,
  ShipMethodID INT NOT NULL,
  CreditCardID INT NULL,
  CreditCardApprovalCode varchar(15) NULL,   
  CurrencyRateID INT NULL,
  SubTotal numeric NOT NULL CONSTRAINT "DF_SalesOrderHeader_SubTotal" DEFAULT (0.00), -- money
  TaxAmt numeric NOT NULL CONSTRAINT "DF_SalesOrderHeader_TaxAmt" DEFAULT (0.00), -- money
  Freight numeric NOT NULL CONSTRAINT "DF_SalesOrderHeader_Freight" DEFAULT (0.00), -- money
  TotalDue numeric, -- AS ISNULL(SubTotal + TaxAmt + Freight, 0),
  Comment varchar(128) NULL,
  rowguid varchar(50) NOT NULL, -- ROWGUIDCOL
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_SalesOrderHeader_ModifiedDate" DEFAULT (NOW()),
  CONSTRAINT "CK_SalesOrderHeader_Status" CHECK (Status BETWEEN 0 AND 8),
  CONSTRAINT "CK_SalesOrderHeader_DueDate" CHECK (DueDate >= OrderDate),
  CONSTRAINT "CK_SalesOrderHeader_ShipDate" CHECK ((ShipDate >= OrderDate) OR (ShipDate IS NULL)),
  CONSTRAINT "CK_SalesOrderHeader_SubTotal" CHECK (SubTotal >= 0.00),
  CONSTRAINT "CK_SalesOrderHeader_TaxAmt" CHECK (TaxAmt >= 0.00),
  CONSTRAINT "CK_SalesOrderHeader_Freight" CHECK (Freight >= 0.00)
);

CREATE TABLE SalesPerson(
  BusinessEntityID INT NOT NULL,
  TerritoryID INT NULL,
  SalesQuota numeric NULL, -- money
  Bonus numeric NOT NULL CONSTRAINT "DF_SalesPerson_Bonus" DEFAULT (0.00), -- money
  CommissionPct numeric NOT NULL CONSTRAINT "DF_SalesPerson_CommissionPct" DEFAULT (0.00), -- smallmoney -- money
  SalesYTD numeric NOT NULL CONSTRAINT "DF_SalesPerson_SalesYTD" DEFAULT (0.00), -- money
  SalesLastYear numeric NOT NULL CONSTRAINT "DF_SalesPerson_SalesLastYear" DEFAULT (0.00), -- money
  rowguid varchar(50) NOT NULL , -- ROWGUIDCOL
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_SalesPerson_ModifiedDate" DEFAULT (NOW()),
  CONSTRAINT "CK_SalesPerson_SalesQuota" CHECK (SalesQuota > 0.00),
  CONSTRAINT "CK_SalesPerson_Bonus" CHECK (Bonus >= 0.00),
  CONSTRAINT "CK_SalesPerson_CommissionPct" CHECK (CommissionPct >= 0.00),
  CONSTRAINT "CK_SalesPerson_SalesYTD" CHECK (SalesYTD >= 0.00),
  CONSTRAINT "CK_SalesPerson_SalesLastYear" CHECK (SalesLastYear >= 0.00)
);

CREATE TABLE SalesTerritory(
  TerritoryID SERIAL NOT NULL, -- int
  Name varchar(50) NOT NULL,
  CountryRegionCode varchar(3) NOT NULL,
  "group" varchar(50) NOT NULL, -- Group
  SalesYTD numeric NOT NULL CONSTRAINT "DF_SalesTerritory_SalesYTD" DEFAULT (0.00), -- money
  SalesLastYear numeric NOT NULL CONSTRAINT "DF_SalesTerritory_SalesLastYear" DEFAULT (0.00), -- money
  CostYTD numeric NOT NULL CONSTRAINT "DF_SalesTerritory_CostYTD" DEFAULT (0.00), -- money
  CostLastYear numeric NOT NULL CONSTRAINT "DF_SalesTerritory_CostLastYear" DEFAULT (0.00), -- money
  rowguid varchar(50) NOT NULL , -- ROWGUIDCOL
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_SalesTerritory_ModifiedDate" DEFAULT (NOW()),
  CONSTRAINT "CK_SalesTerritory_SalesYTD" CHECK (SalesYTD >= 0.00),
  CONSTRAINT "CK_SalesTerritory_SalesLastYear" CHECK (SalesLastYear >= 0.00),
  CONSTRAINT "CK_SalesTerritory_CostYTD" CHECK (CostYTD >= 0.00),
  CONSTRAINT "CK_SalesTerritory_CostLastYear" CHECK (CostLastYear >= 0.00)
);

CREATE TABLE Currency(
  CurrencyCode char(3) NOT NULL,
  Name varchar(50) NOT NULL,
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_Currency_ModifiedDate" DEFAULT (NOW())
);
CREATE TABLE CurrencyRate(
  CurrencyRateID SERIAL NOT NULL, -- int
  CurrencyRateDate TIMESTAMP NOT NULL,   
  FromCurrencyCode char(3) NOT NULL,
  ToCurrencyCode char(3) NOT NULL,
  AverageRate numeric NOT NULL, -- money
  EndOfDayRate numeric NOT NULL,  -- money
  ModifiedDate TIMESTAMP NOT NULL CONSTRAINT "DF_CurrencyRate_ModifiedDate" DEFAULT (NOW())
);

INSERT INTO ProductReview (ProductReviewID, ProductID, ReviewerName, ReviewDate, EmailAddress, Rating, Comments, ModifiedDate) VALUES
 (1, 709, 'John Smith', '2013-09-18 00:00:00', 'john@fourthcoffee.com', 5, 'I can''t believe I''m singing the praises of a pair of socks, but I just came back from a grueling
3-day ride and these socks really helped make the trip a blast. They''re lightweight yet really cushioned my feet all day. 
The reinforced toe is nearly bullet-proof and I didn''t experience any problems with rubbing or blisters like I have with
other brands. I know it sounds silly, but it''s always the little stuff (like comfortable feet) that makes or breaks a long trip.
I won''t go on another trip without them!', '2013-09-18 00:00:00'),

 (2, 937, 'David', '2013-11-13 00:00:00', 'david@graphicdesigninstitute.com', 4, 'A little on the heavy side, but overall the entry/exit is easy in all conditions. I''ve used these pedals for 
more than 3 years and I''ve never had a problem. Cleanup is easy. Mud and sand don''t get trapped. I would like 
them even better if there was a weight reduction. Maybe in the next design. Still, I would recommend them to a friend.', '2013-11-13 00:00:00'),

 (3, 937, 'Jill', '2013-11-15 00:00:00', 'jill@margiestravel.com', 2, 'Maybe it''s just because I''m new to mountain biking, but I had a terrible time getting use
to these pedals. In my first outing, I wiped out trying to release my foot. Any suggestions on
ways I can adjust the pedals, or is it just a learning curve thing?', '2013-11-15 00:00:00'),

 (4, 798, 'Laura Norman', '2013-11-15 00:00:00', 'laura@treyresearch.net', 5, 'The Road-550-W from Adventure Works Cycles is everything it''s advertised to be. Finally, a quality bike that
is actually built for a woman and provides control and comfort in one neat package. The top tube is shorter, the suspension is weight-tuned and there''s a much shorter reach to the brake
levers. All this adds up to a great mountain bike that is sure to accommodate any woman''s anatomy. In addition to getting the size right, the saddle is incredibly comfortable. 
Attention to detail is apparent in every aspect from the frame finish to the careful design of each component. Each component is a solid performer without any fluff. 
The designers clearly did their homework and thought about size, weight, and funtionality throughout. And at less than 19 pounds, the bike is manageable for even the most petite cyclist.

We had 5 riders take the bike out for a spin and really put it to the test. The results were consistent and very positive. Our testers loved the manuverability 
and control they had with the redesigned frame on the 550-W. A definite improvement over the 2012 design. Four out of five testers listed quick handling
and responsivness were the key elements they noticed. Technical climbing and on the flats, the bike just cruises through the rough. Tight corners and obstacles were handled effortlessly. The fifth tester was more impressed with the smooth ride. The heavy-duty shocks absorbed even the worst bumps and provided a soft ride on all but the 
nastiest trails and biggest drops. The shifting was rated superb and typical of what we''ve come to expect from Adventure Works Cycles. On descents, the bike handled flawlessly and tracked very well. The bike is well balanced front-to-rear and frame flex was minimal. In particular, the testers
noted that the brake system had a unique combination of power and modulation.  While some brake setups can be overly touchy, these brakes had a good
amount of power, but also a good feel that allows you to apply as little or as much braking power as is needed. Second is their short break-in period. We found that they tend to break-in well before
the end of the first ride; while others take two to three rides (or more) to come to full power. 

On the negative side, the pedals were not quite up to our tester''s standards. 
Just for fun, we experimented with routine maintenance tasks. Overall we found most operations to be straight forward and easy to complete. The only exception was replacing the front wheel. The maintenance manual that comes
with the bike say to install the front wheel with the axle quick release or bolt, then compress the fork a few times before fastening and tightening the two quick-release mechanisms on the bottom of the dropouts. This is to seat the axle in the dropouts, and if you do not
do this, the axle will become seated after you tightened the two bottom quick releases, which will then become loose. It''s better to test the tightness carefully or you may notice that the two bottom quick releases have come loose enough to fall completely open. And that''s something you don''t want to experience
while out on the road! 

The Road-550-W frame is available in a variety of sizes and colors and has the same durable, high-quality aluminum that AWC is known for. At a MSRP of just under $1125.00, it''s comparable in price to its closest competitors and
we think that after a test drive you''l find the quality and performance above and beyond . You''ll have a grin on your face and be itching to get out on the road for more. While designed for serious road racing, the Road-550-W would be an excellent choice for just about any terrain and 
any level of experience. It''s a huge step in the right direction for female cyclists and well worth your consideration and hard-earned money.', '2013-11-15 00:00:00');

\copy BusinessEntity FROM './data/BusinessEntity.csv' DELIMITER E'\t' CSV;^X
\copy Person FROM './data/Person.csv' DELIMITER E'\t' CSV;
\copy ProductSubcategory FROM './data/ProductSubcategory.csv' DELIMITER E'\t' CSV;
\copy ProductModel FROM './data/ProductModel.csv' DELIMITER E'\t' CSV;
\copy Product FROM './data/Product.csv' DELIMITER E'\t' CSV;
\copy ProductModelProductDescriptionCulture FROM './data/ProductModelProductDescriptionCulture.csv' DELIMITER E'\t' CSV;
\copy CountryRegionCurrency FROM './data/CountryRegionCurrency.csv' DELIMITER E'\t' CSV;
\copy Currency FROM './data/Currency.csv' DELIMITER E'\t' CSV;
\copy CurrencyRate FROM './data/CurrencyRate.csv' DELIMITER E'\t' CSV;
\copy SalesOrderHeader FROM './data/SalesOrderHeader.csv' DELIMITER E'\t' CSV;
\copy SalesPerson FROM './data/SalesPerson.csv' DELIMITER E'\t' CSV;
\copy SalesTerritory FROM './data/SalesTerritory.csv' DELIMITER E'\t' CSV;


**Answer.**

---------

## Finding our most popular products

As discussed, the company would like to know which of their products is the most popular among customers. You figure that the average rating given in reviews is correlated with the number of sales of a particular product (that products with higher reviews have more sales).

### Exercise 2:

Using the ```product``` and ```productreview``` tables, ```JOIN``` them and rank the products according to their average review rating. What are the names and IDs of the top 5 products?

**Query**

SELECT prod.Name as Product_Name, AVG(prodrev.Rating) as Avg_rating, prod.productID as Prod_ID FROM Product Prod JOIN ProductReview ProdRev ON Prod.ProductID = ProdRev.ProductID GROUP BY prod.name, prod.productid ORDER BY Avg_rating DESC;

In [74]:
%%sql
SELECT prod.Name as Product_Name, AVG(prodrev.Rating) as Avg_rating, prod.productID as Prod_ID FROM Product Prod JOIN ProductReview ProdRev ON Prod.ProductID = ProdRev.ProductID GROUP BY prod.name, prod.productid ORDER BY Avg_rating DESC;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
3 rows affected.


[('Mountain Bike Socks, M', Decimal('5.0000000000000000'), 709),
 ('Road-550-W Yellow, 40', Decimal('5.0000000000000000'), 798),
 ('HL Mountain Pedal', Decimal('3.0000000000000000'), 937)]

---------

### Exercise 3:

Much to your disappointment, there are only three products with ratings and only four reviews in total! This is nowhere near enough to perform an analysis of the correlation between reviews and total sales.

Nevertheless, your manager wants the **English description** of these products for an upcoming sale. Use the documentation provided above if you need help navigating the structure to extract this!

**Hint:** You'll notice that the value for `cultureid` in the `productmodelproductdescriptionculture` table often has extra trailing spaces which makes it difficult to reliably get descriptions of a specific language. You should first modify this table before writing the `SELECT` statement to get the descriptions that your manager wants. To do this, you can use an `UPDATE` statement with Postgres's [`TRIM`](https://w3resource.com/PostgreSQL/trim-function.php) function.

**Query**

SET client_encoding TO 'UTF8';

\copy ProductModelProductDescriptionCulture FROM './data/ProductModelProductDescriptionCulture.csv' DELIMITER E'\t' CSV;

UPDATE productmodelproductdescriptionculture SET CultureID = TRIM(from cultureid);


WITH top_products AS ( SELECT prod.Name as Product_Name, AVG(prodrev.Rating) as Avg_rating, prod.productID as Prod_ID, prod.productmodelid as Product_Model_ID FROM Product Prod JOIN ProductReview ProdRev ON Prod.ProductID = ProdRev.ProductID GROUP BY prod.name, prod.productid, prod.productmodelid ORDER BY Avg_rating DESC),
desc_ AS ( SELECT proddesc.Description as Prod_Description, desclan.cultureid as Description_Language, desclan.productmodelid as Product_Model_ID FROM ProductDescription proddesc JOIN ProductModelProductDescriptionCulture desclan ON proddesc.ProductDescriptionID = desclan.ProductDescriptionID WHERE cultureid = 'en') 
SELECT Product_Name, Prod_Description, Prod_ID FROM top_products JOIN desc_ ON top_products.Product_Model_ID = desc_.Product_Model_ID;


In [76]:
%%sql
WITH top_products AS ( SELECT prod.Name as Product_Name, AVG(prodrev.Rating) as Avg_rating, prod.productID as Prod_ID, prod.productmodelid as Product_Model_ID FROM Product Prod JOIN ProductReview ProdRev ON Prod.ProductID = ProdRev.ProductID GROUP BY prod.name, prod.productid, prod.productmodelid ORDER BY Avg_rating DESC),
desc_ AS ( SELECT proddesc.Description as Prod_Description, desclan.cultureid as Description_Language, desclan.productmodelid as Product_Model_ID FROM ProductDescription proddesc JOIN ProductModelProductDescriptionCulture desclan ON proddesc.ProductDescriptionID = desclan.ProductDescriptionID WHERE cultureid = 'en') 
SELECT Product_Name, Prod_Description, Prod_ID FROM top_products JOIN desc_ ON top_products.Product_Model_ID = desc_.Product_Model_ID;


 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
3 rows affected.


[('Road-550-W Yellow, 40', 'Same technology as all of our Road series bikes, but the frame is sized for a woman.  Perfect all-around bike for road or racing.', 798),
 ('HL Mountain Pedal', 'Stainless steel; designed to shed mud easily.', 937),
 ('Mountain Bike Socks, M', 'Combination of natural and synthetic fibers stays dry and provides just the right cushioning.', 709)]

---------

### Exercise 4:

Since we cannot infer the most popular products from the reviews, we will go with an alternative strategy.

Get the model ID, name, description, and total number of sales for each product and display the top-10 selling products. You can infer how often products have been sold by looking at the `salesorderdetail` table (each row might indicate more than one sale, so take note of `OrderQty`).

**Query**

SELECT ProductID, SUM(OrderQty) as Total_sold FROM salesorderdetail GROUP BY ProductID ORDER BY Total_sold DESC;

WITH top_products AS ( SELECT prod.Name as Product_Name, SUM(OrderQty) as Total_sold, prod.productID as Prod_ID, prod.productmodelid as Product_Model_ID FROM Product Prod JOIN salesorderdetail SO_detail ON Prod.ProductID = SO_detail.ProductID GROUP BY prod.name, prod.productid, prod.productmodelid ORDER BY Total_sold DESC),
desc_ AS ( SELECT proddesc.Description as Prod_Description, desclan.cultureid as Description_Language, desclan.productmodelid as Product_Model_ID FROM ProductDescription proddesc JOIN ProductModelProductDescriptionCulture desclan ON proddesc.ProductDescriptionID = desclan.ProductDescriptionID WHERE cultureid = 'en') 
SELECT Product_Name, top_products.Product_Model_ID, Prod_Description, Total_sold FROM top_products JOIN desc_ ON top_products.Product_Model_ID = desc_.Product_Model_ID LIMIT 10;



**Answer.**

In [77]:
%%sql
WITH top_products AS ( SELECT prod.Name as Product_Name, SUM(OrderQty) as Total_sold, prod.productID as Prod_ID, prod.productmodelid as Product_Model_ID FROM Product Prod JOIN salesorderdetail SO_detail ON Prod.ProductID = SO_detail.ProductID GROUP BY prod.name, prod.productid, prod.productmodelid ORDER BY Total_sold DESC),
desc_ AS ( SELECT proddesc.Description as Prod_Description, desclan.cultureid as Description_Language, desclan.productmodelid as Product_Model_ID FROM ProductDescription proddesc JOIN ProductModelProductDescriptionCulture desclan ON proddesc.ProductDescriptionID = desclan.ProductDescriptionID WHERE cultureid = 'en') 
SELECT Product_Name, top_products.Product_Model_ID, Prod_Description, Total_sold FROM top_products JOIN desc_ ON top_products.Product_Model_ID = desc_.Product_Model_ID LIMIT 10;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
10 rows affected.


[('AWC Logo Cap', 2, 'Traditional style with a flip-up brim; one-size fits all.', 8311),
 ('Water Bottle - 30 oz.', 111, 'AWC logo water bottle - holds 30 oz; leak-proof.', 6815),
 ('Sport-100 Helmet, Blue', 33, 'Universal fit, well-vented, lightweight , snap-on visor.', 6743),
 ('Long-Sleeve Logo Jersey, L', 11, 'Unisex long-sleeve AWC logo microfiber cycling jersey', 6592),
 ('Sport-100 Helmet, Black', 33, 'Universal fit, well-vented, lightweight , snap-on visor.', 6532),
 ('Sport-100 Helmet, Red', 33, 'Universal fit, well-vented, lightweight , snap-on visor.', 6266),
 ('Classic Vest, S', 1, 'Light-weight, wind-resistant, packs to fit into a pocket.', 4247),
 ('Patch Kit/8 Patches', 114, 'Includes 8 different size patches, glue and sandpaper.', 3865),
 ('Short-Sleeve Classic Jersey, XL', 32, 'Short sleeve classic breathable jersey with superior moisture control, front zipper, and 3 back pockets.', 3864),
 ('Long-Sleeve Logo Jersey, M', 11, 'Unisex long-sleeve AWC logo microfiber cycling jersey', 3636)]

---------

### Exercise 5:

Let's look at the correlation between quantity sold and price for each item in each subcategory. Some subcategories don't have enough sales to make the correlation meaningful, so only look at the top 10 subcategories by total quantity of sales.

Once you've looked at the data, make a hypothesis about what causes any positive or negative correlations between price and quantity, and explain this in 2-3 sentences.

**Hint:** You'll need to calculate the total quantities from `salesorderdetail` again and group the products by subcategory. It'll probably be easier if you use at least two [CTEs](https://www.postgresql.org/docs/9.1/queries-with.html). You can calculate the correlation in PostgreSQL by using the built-in [```corr()```](https://www.postgresql.org/docs/9.4/functions-aggregate.html) function.

**Query**

SELECT ProductID, SUM(OrderQty) as Total_sold FROM salesorderdetail JOIN GROUP BY ProductID ORDER BY Total_sold DESC;

This was done to delete duplicates
DELETE
FROM
    ProductModelProductDescriptionCulture a
        USING ProductModelProductDescriptionCulture b
WHERE
    a.productmodelid = b.productmodelid;
    
SELECT ProductID, SUM(OrderQty) as Total_sold FROM salesorderdetail GROUP BY ProductID ORDER BY Total_sold DESC;

WITH sales_subcat AS ( SELECT prod.ProductSubcategoryID as Product_subcateg,  SUM(OrderQty) as Total_sold, AVG(ListPrice) as Average_price, Corr(OrderQty,ListPrice) as Correlation FROM Product Prod JOIN salesorderdetail SO_detail ON Prod.ProductID = SO_detail.ProductID GROUP BY prod.ProductSubcategoryID  ORDER BY Total_sold DESC),

sales_subcat_name AS ( SELECT sales_subcat.Product_subcateg as Product_subcateg, ProductSubcategory.Name as Subcategory_Name, sales_subcat.Total_sold as Total_sold, sales_subcat.Average_price as Average_price, sales_subcat.Correlation FROM sales_subcat JOIN ProductSubcategory ON sales_subcat.Product_subcateg = ProductSubcategory.ProductSubcategoryID ORDER BY Total_sold DESC) 
SELECT * FROM sales_subcat_name limit 10;

In [78]:
%%sql

WITH sales_subcat AS ( SELECT prod.ProductSubcategoryID as Product_subcateg,  SUM(OrderQty) as Total_sold, AVG(ListPrice) as Average_price, Corr(OrderQty,ListPrice) as Correlation FROM Product Prod JOIN salesorderdetail SO_detail ON Prod.ProductID = SO_detail.ProductID GROUP BY prod.ProductSubcategoryID  ORDER BY Total_sold DESC),

sales_subcat_name AS ( SELECT sales_subcat.Product_subcateg as Product_subcateg, ProductSubcategory.Name as Subcategory_Name, sales_subcat.Total_sold as Total_sold, sales_subcat.Average_price as Average_price, sales_subcat.Correlation FROM sales_subcat JOIN ProductSubcategory ON sales_subcat.Product_subcateg = ProductSubcategory.ProductSubcategoryID ORDER BY Total_sold DESC) 
SELECT * FROM sales_subcat_name limit 10;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
10 rows affected.


[(2, 'Road Bikes', 47196, Decimal('1529.7549713521772345'), -0.1597486954281),
 (1, 'Mountain Bikes', 28321, Decimal('1968.6853520109175564'), 0.0658142818788795),
 (21, 'Jerseys', 22711, Decimal('51.5658587122792379'), 0.0107191293480794),
 (31, 'Helmets', 19541, Decimal('34.9900000000000000'), None),
 (37, 'Tires and Tubes', 18006, Decimal('14.0555924549871392'), -0.0721532678808808),
 (3, 'Touring Bikes', 14751, Decimal('1566.2765610859728507'), -0.041872718132813),
 (20, 'Gloves', 13012, Decimal('28.0962413314840499'), 0.383723703588632),
 (14, 'Road Frames', 11753, Decimal('586.1204646721833227'), -0.0520611587216979),
 (12, 'Mountain Frames', 11621, Decimal('752.4969510833147197'), -0.140263296256509),
 (28, 'Bottles and Cages', 10552, Decimal('7.0045994065281899'), -0.155751654618702)]

**ANSWER**

In general, we can see a low correlation coefficient value, which most likely implies that the relationship between the listed price and the sold quantity is not very linear.

Also, there is a negative correlation in most of the subcategories and still very low. There might be more information inside each subcategory, meaning that some products might have a positive while others a negative and we are only getting an "average" result that does not allow to draw many conclusions.

If we do not consider the previous idea, we can see that gloves, jerseys and mountain bikes have a positive correlation, which could be an indicator that the customers are more whiling to pay the listed price for these products. While the other elements have a negative correlation, meaning that probably these products are less bought when their prices increase. It is also noticeable that there is no correlation for helmets meaning that probably the sold quantity is independent of price.

---------

## Finding our top salespeople

As mentioned earlier, we want to find our best salespeople and see whether or not we can incentivize them in an appropriate manner. Namely, we want to determine if the commission percentage we give them motivates them to make more and bigger sales.

### Exercise 5:

Find the top five performing salespeople by using the `salesytd` (Sales, year-to-date) column. (We only need to know the `businessentityid` for each salesperson as this uniquely identifies each.) Why might you be skeptical of these numbers right now?

**Query**

SELECT SalesPerson.BusinessEntityID, SalesPerson.SalesYTD, SalesPerson.SalesQuota, SalesPerson.SalesLastYear FROM SalesPerson ORDER BY SalesYTD DESC LIMIT 5;

In [79]:
%%sql
SELECT SalesPerson.BusinessEntityID, SalesPerson.SalesYTD, SalesPerson.SalesQuota, SalesPerson.SalesLastYear FROM SalesPerson ORDER BY SalesYTD DESC LIMIT 5;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
5 rows affected.


[(276, Decimal('4251368.5497'), Decimal('250000'), Decimal('1439156.0291')),
 (289, Decimal('4116871.2277'), Decimal('250000'), Decimal('1635823.3967')),
 (275, Decimal('3763178.1787'), Decimal('300000'), Decimal('1750406.4785')),
 (277, Decimal('3189418.3662'), Decimal('250000'), Decimal('1997186.2037')),
 (290, Decimal('3121616.3202'), Decimal('250000'), Decimal('2396539.7601'))]

**Answer.**

Looking at the salesytd column, we can see that the top performers are selling over 10 times their sales cuota and it is not clear what percentage of the fiscal year has passed at the moment the data was taken.

Also, all the sales from last year are way below the salesytd (which does not comprise the sales of a whole year), therefore, the data seems suspicious and more information should be taken into account.

---------

### Exercise 6:

Using ```salesorderheader```, find the top 5 salespeople who made the most sales **in the most recent year** (2014). (There is a column called `subtotal` - use that.) Sales that do not have an associated salesperson should be excluded from your calculations and final output. All orders that were made within the 2014 calendar year should be included.

**Hint:** You can use the syntax `'1970-01-01'::date` to generate an arbitrary date in PostgreSQL and compare this to specific dates in the tables.

**Query**

SELECT salesorderheader.salespersonid, sum(salesorderheader.subtotal) as Sales_2014 FROM salesorderheader WHERE salespersonid is not NULL AND orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salespersonid ORDER BY Sales_2014 DESC LIMIT 5;

In [80]:
%%sql
SELECT salesorderheader.salespersonid, sum(salesorderheader.subtotal) as Sales_2014 FROM salesorderheader WHERE salespersonid is not NULL AND orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salespersonid ORDER BY Sales_2014 DESC LIMIT 5;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
5 rows affected.


[(289, Decimal('1382996.5839')),
 (276, Decimal('1271088.5216')),
 (275, Decimal('1057247.3786')),
 (282, Decimal('1044810.8277')),
 (277, Decimal('1040093.4071'))]

---------

### Exercise 7:

Looking at the documentation, you will see that `subtotal` in the ```salesorderheader``` table is calculated from other tables in the database. To validate this figure (instead of trusting it blindly), let's calculate `subtotal` manually. Using the ```salesorderdetail``` and ```salesorderheader``` tables, calculate the sales for each salesperson for **this past year** (2014) and display results for the top 5 salespeople.

**Hint:** You will have to ```JOIN``` ```salesorderdetail``` on ```salesorderheader``` to get the salesperson, calculate line totals for each sale using appropriate discounts, then sum all the line totals to get the total sale. You will want to use ```WITH``` clauses again to keep things sane.

**Query**

WITH sales_header AS (SELECT salesorderheader.salespersonID, SUM(salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount)) as Total_sales_2014 FROM salesorderheader JOIN salesorderdetail ON salesorderheader.salesorderID = salesorderdetail.salesorderID WHERE salesorderheader.salespersonID is not NULL AND salesorderheader.orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salesorderheader.salespersonid ORDER BY Total_sales_2014 DESC)
SELECT * FROM sales_header LIMIT 5;

In [81]:
%%sql
WITH sales_header AS (SELECT salesorderheader.salespersonID, SUM(salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount)) as Total_sales_2014 FROM salesorderheader JOIN salesorderdetail ON salesorderheader.salesorderID = salesorderdetail.salesorderID WHERE salesorderheader.salespersonID is not NULL AND salesorderheader.orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salesorderheader.salespersonid ORDER BY Total_sales_2014 DESC)
SELECT * FROM sales_header LIMIT 5;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
5 rows affected.


[(289, Decimal('1382996.583910')),
 (276, Decimal('1271088.521461')),
 (275, Decimal('1057247.378572')),
 (282, Decimal('1044810.827687')),
 (277, Decimal('1040093.406901'))]

---------

### Exercise 8:

Using ```corr()```, see if there is a positive relationship between total sales and commission percentage.

**Query**

WITH sales_commi AS (SELECT salesorderheader.salespersonID as Salesperson_ID, SUM(SalesOrderHeader.SubTotal) as Total_sales_2014, AVG(salesperson.CommissionPct) as Commission FROM salesorderheader JOIN salesperson ON salesorderheader.salespersonID = salesperson.businessentityID WHERE salesorderheader.salespersonID is not NULL AND salesorderheader.orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salesorderheader.salespersonid)
SELECT CORR(Total_sales_2014, Commission) as Correlation FROM sales_commi;

In [82]:
%%sql

WITH sales_commi AS (SELECT salesorderheader.salespersonID as Salesperson_ID, SUM(SalesOrderHeader.SubTotal) as Total_sales_2014, AVG(salesperson.CommissionPct) as Commission FROM salesorderheader JOIN salesperson ON salesorderheader.salespersonID = salesperson.businessentityID WHERE salesorderheader.salespersonID is not NULL AND salesorderheader.orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salesorderheader.salespersonid)
SELECT CORR(Total_sales_2014, Commission) as Correlation FROM sales_commi;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
1 rows affected.


[(0.737527789212099,)]

**Answer.**

When calculating the correlation between total sales in 2014 for all the salesperson and their average commissions per sale, we can see a strong correlation. Which implies a, on one side, linear relationship between total sales and commission and, on the other side, that more sales imply more commission and viceversa (as expected).

---------

### Exercise 9:

Remember how we mentioned that products were sold in many regions? This is why you had to work with the `culture` value before to get the English language descriptions. To make matters worse, you are told the sales are recorded in **local** currency, so your previous analysis is flawed, and you must convert all amounts to USD if you wish to compare the different salespeople fairly!

Use the `countryregioncurrency` table in combination with the `salesperson` and `salesterritory` ones to figure out the relevant currency symbol for each of the top salespeople.

**Query**

select * from countryregioncurrency limit 10;

select * from salesterritory limit 10;

SELECT salesperson.businessentityID as SalesPersonID, countryregioncurrency.currencycode as Currency FROM SalesPerson
JOIN SalesTerritory ON SalesPerson.TerritoryID = SalesTerritory.TerritoryID
JOIN countryregioncurrency ON SalesTerritory.Countryregioncode = countryregioncurrency.Countryregioncode;

In [83]:
%%sql

SELECT salesperson.businessentityID as SalesPersonID, countryregioncurrency.currencycode as Currency FROM SalesPerson
JOIN SalesTerritory ON SalesPerson.TerritoryID = SalesTerritory.TerritoryID
JOIN countryregioncurrency ON SalesTerritory.Countryregioncode = countryregioncurrency.Countryregioncode;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
16 rows affected.


[(275, 'USD'),
 (276, 'USD'),
 (277, 'USD'),
 (278, 'CAD'),
 (279, 'USD'),
 (280, 'USD'),
 (281, 'USD'),
 (282, 'CAD'),
 (283, 'USD'),
 (284, 'USD'),
 (286, 'AUD'),
 (288, 'DEM'),
 (288, 'EUR'),
 (289, 'GBP'),
 (290, 'EUR'),
 (290, 'FRF')]

---------

### Exercise 10:

Now that we have the currency codes associated with each salesperson, redo Exercise 7 to take the currency exchange into account. If there are salespeople in the top 5 that weren't there before, explain why.

**Hint:** The rates in the```currencyrate``` table always go from `FromCurrencyCode=USD` to `ToCurrencyCode=<Desired Currency Code>`, and they are listed every day. When calculating line totals, use the `AverageRate` for that day. You should be able to reuse a lot of Exercise 7.

**Query**

WITH sales_header AS (SELECT salesorderheader.salespersonID, 
SUM( CASE WHEN countryregioncurrency.currencycode = 'USD' THEN
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount) 
ELSE 
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount)*currencyrate.averagerate 
END) as Total_sales_2014_converted_currency,
countryregioncurrency.currencycode as Original_Currency
FROM salesorderheader 
JOIN salesorderdetail ON salesorderheader.salesorderID = salesorderdetail.salesorderID
JOIN currencyrate ON salesorderheader.orderdate = currencyrate.currencyratedate
JOIN SalesPerson ON salesorderheader.salespersonID = salesperson.businessentityID
JOIN SalesTerritory ON SalesPerson.TerritoryID = SalesTerritory.TerritoryID
JOIN countryregioncurrency ON SalesTerritory.Countryregioncode = countryregioncurrency.Countryregioncode
WHERE salesorderheader.salespersonID is not NULL AND salesorderheader.orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salesorderheader.salespersonid, countryregioncurrency.currencycode ORDER BY Total_sales_2014_converted_currency DESC)
SELECT * FROM sales_header ORDER BY Total_sales_2014_converted_currency DESC;


In [84]:
%%sql

WITH sales_header AS (SELECT salesorderheader.salespersonID, 
SUM( CASE WHEN countryregioncurrency.currencycode = 'USD' THEN
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount) 
ELSE 
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount)*currencyrate.averagerate 
END) as Total_sales_2014_converted_currency,
countryregioncurrency.currencycode as Original_Currency
FROM salesorderheader 
JOIN salesorderdetail ON salesorderheader.salesorderID = salesorderdetail.salesorderID
JOIN currencyrate ON salesorderheader.orderdate = currencyrate.currencyratedate
JOIN SalesPerson ON salesorderheader.salespersonID = salesperson.businessentityID
JOIN SalesTerritory ON SalesPerson.TerritoryID = SalesTerritory.TerritoryID
JOIN countryregioncurrency ON SalesTerritory.Countryregioncode = countryregioncurrency.Countryregioncode
WHERE salesorderheader.salespersonID is not NULL AND salesorderheader.orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salesorderheader.salespersonid, countryregioncurrency.currencycode ORDER BY Total_sales_2014_converted_currency DESC)
SELECT * FROM sales_header ORDER BY Total_sales_2014_converted_currency DESC;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
16 rows affected.


[(289, Decimal('2128299780.6674506459'), 'GBP'),
 (282, Decimal('1614351721.0866827115'), 'CAD'),
 (290, Decimal('1375205450.8460989870'), 'FRF'),
 (290, Decimal('1375205450.8460989870'), 'EUR'),
 (288, Decimal('905955605.7368302157'), 'EUR'),
 (288, Decimal('905955605.7368302157'), 'DEM'),
 (286, Decimal('901383304.0419726912'), 'AUD'),
 (278, Decimal('680889921.3876525586'), 'CAD'),
 (276, Decimal('15253062.257532'), 'USD'),
 (275, Decimal('12686968.542864'), 'USD'),
 (277, Decimal('12481120.882812'), 'USD'),
 (279, Decimal('9446453.146080'), 'USD'),
 (281, Decimal('9335299.824276'), 'USD'),
 (284, Decimal('7211966.044296'), 'USD'),
 (280, Decimal('6059184.528480'), 'USD'),
 (283, Decimal('5885595.826980'), 'USD')]

**Query**

select * from currencyrate where tocurrencycode = 'AUD' limit 10;

WITH sales_header AS (SELECT salesorderheader.salespersonID, 
SUM( CASE WHEN countryregioncurrency.currencycode = 'USD' THEN
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount) 
ELSE 
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount)*currencyrate.averagerate 
END) as Total_sales_2014_converted_currency,
countryregioncurrency.currencycode as Original_Currency
FROM salesorderheader 
JOIN salesorderdetail ON salesorderheader.salesorderID = salesorderdetail.salesorderID
JOIN currencyrate ON salesorderheader.orderdate = currencyrate.currencyratedate
JOIN SalesPerson ON salesorderheader.salespersonID = salesperson.businessentityID
JOIN SalesTerritory ON SalesPerson.TerritoryID = SalesTerritory.TerritoryID
JOIN countryregioncurrency ON SalesTerritory.Countryregioncode = countryregioncurrency.Countryregioncode
WHERE salesorderheader.salespersonID is not NULL AND salesorderheader.orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salesorderheader.salespersonid, countryregioncurrency.currencycode ORDER BY Total_sales_2014_converted_currency DESC)
SELECT salespersonid, SUM(Total_sales_2014_converted_currency) AS Total_sales_2014 FROM sales_header GROUP BY salespersonid ORDER BY Total_sales_2014 DESC LIMIT 5;


In [85]:
%%sql

WITH sales_header AS (SELECT salesorderheader.salespersonID, 
SUM( CASE WHEN countryregioncurrency.currencycode = 'USD' THEN
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount) 
ELSE 
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount)*currencyrate.averagerate 
END) as Total_sales_2014_converted_currency,
countryregioncurrency.currencycode as Original_Currency
FROM salesorderheader 
JOIN salesorderdetail ON salesorderheader.salesorderID = salesorderdetail.salesorderID
JOIN currencyrate ON salesorderheader.orderdate = currencyrate.currencyratedate
JOIN SalesPerson ON salesorderheader.salespersonID = salesperson.businessentityID
JOIN SalesTerritory ON SalesPerson.TerritoryID = SalesTerritory.TerritoryID
JOIN countryregioncurrency ON SalesTerritory.Countryregioncode = countryregioncurrency.Countryregioncode
WHERE salesorderheader.salespersonID is not NULL AND salesorderheader.orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salesorderheader.salespersonid, countryregioncurrency.currencycode ORDER BY Total_sales_2014_converted_currency DESC)
SELECT salespersonid, SUM(Total_sales_2014_converted_currency) AS Total_sales_2014 FROM sales_header GROUP BY salespersonid ORDER BY Total_sales_2014 DESC LIMIT 5;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
5 rows affected.


[(290, Decimal('2750410901.6921979740')),
 (289, Decimal('2128299780.6674506459')),
 (288, Decimal('1811911211.4736604314')),
 (282, Decimal('1614351721.0866827115')),
 (286, Decimal('901383304.0419726912'))]

**Answer.**
           
From the initial exploration in this exercise, we could see that a same sales person could have businesses with different currencies. Therefore, their total sales were not comparable. When they were all converted and added, the total was very different compared to exercise 7. In fact only the sales persons with ID 289 and 282 remained in the top 5.

---------

### Exercise 11:

How does the correlation from Exercise 8 change once you've adjusted for the currency?

**Query**

WITH sales_header AS (SELECT salesorderheader.salespersonID, 
SUM( CASE WHEN countryregioncurrency.currencycode = 'USD' THEN
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount) 
ELSE 
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount)*currencyrate.averagerate 
END) as Total_sales_2014,
AVG(salesperson.CommissionPct) as Commission
FROM salesorderheader 
JOIN salesorderdetail ON salesorderheader.salesorderID = salesorderdetail.salesorderID
JOIN currencyrate ON salesorderheader.orderdate = currencyrate.currencyratedate
JOIN SalesPerson ON salesorderheader.salespersonID = salesperson.businessentityID
JOIN SalesTerritory ON SalesPerson.TerritoryID = SalesTerritory.TerritoryID
JOIN countryregioncurrency ON SalesTerritory.Countryregioncode = countryregioncurrency.Countryregioncode
WHERE salesorderheader.salespersonID is not NULL AND salesorderheader.orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salesorderheader.salespersonid ORDER BY Total_sales_2014 DESC)
SELECT CORR(Total_sales_2014, Commission) as Correlation FROM sales_header;

In [86]:
%%sql

WITH sales_header AS (SELECT salesorderheader.salespersonID, 
SUM( CASE WHEN countryregioncurrency.currencycode = 'USD' THEN
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount) 
ELSE 
salesorderdetail.orderqty*salesorderdetail.unitprice*(1 - salesorderdetail.unitpricediscount)*currencyrate.averagerate 
END) as Total_sales_2014,
AVG(salesperson.CommissionPct) as Commission
FROM salesorderheader 
JOIN salesorderdetail ON salesorderheader.salesorderID = salesorderdetail.salesorderID
JOIN currencyrate ON salesorderheader.orderdate = currencyrate.currencyratedate
JOIN SalesPerson ON salesorderheader.salespersonID = salesperson.businessentityID
JOIN SalesTerritory ON SalesPerson.TerritoryID = SalesTerritory.TerritoryID
JOIN countryregioncurrency ON SalesTerritory.Countryregioncode = countryregioncurrency.Countryregioncode
WHERE salesorderheader.salespersonID is not NULL AND salesorderheader.orderdate BETWEEN '2014-01-01' AND '2014-12-31' GROUP BY salesorderheader.salespersonid ORDER BY Total_sales_2014 DESC)
SELECT CORR(Total_sales_2014, Commission) as Correlation FROM sales_header;

 * postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/adventureworks
   postgresql://postgres:***@extended-case-3.cwoudog5zurm.us-east-2.rds.amazonaws.com/postgres
1 rows affected.


[(0.550397293992879,)]

**Answer.**
 
We can see quite a lower correlation in this case compared to exercise 8 (from ~0.73 down to ~0.55). When considering the currency change, this factor can account for the difficulty or complexity of projects/sales in different regions. 

---------